### ISSUE TITLE
IO Thread Purge policy

### ISSUE THREAD

**Issue by pp112018** [ReactiveX/RxJava#6315]
Hi,

we use RX 1.2.7 and we are trying to debug an issue where we see an abnormally high count of idle RxIoScheduler threads (up to 16k) in some boxes even though traffic type and load to all the boxes are evenly distributed. The variance in those boxes can range from hundreds to the abovementioned 16k.

Is this a normal thing to observe for the CachedThreadScheduler in 1.2.7? We don't use 2.x but is this also a common thing for IOScheduler in 2.x?

We have tried different things to troubleshoot the root cause.

1) Shortening the keep alive time of the ThreadWorkers in CachedThreadScheduler gives us mixed results and we are concerned that too aggressively evicting ThreadWorkers could be counterproductive since starting new ThreadWorkers should be more expensive than reusing existing ones in as much as possible, right? We don't want to solve a performance problem by creating another one. More importantly, since production traffic varies, there's a big question mark on what the optimal keep alive time should be and how could we could be forced to keep adjusting the keep alive time frequently. We had hoped that the IO scheduler thread pool should elastically grow and shrink automatically by itself.

2) Another thought is to upgrade to 2.x where we noticed that the purge thread is turned on by default. 
What is the correct way to purge idle IO scheduler threads in 1.2.7 or 1.x? We are wondering if perhaps some code in our application caused the ThreadWorker lifecycle to not terminate correctly and thus hold on too longer than was necessary but we are not sure what. One potential suspect is the onCompleted call. We noticed that different types of clean up happens in onCompleted in 1.x and onComplete in 2.x.  However, we also noticed that the IO scheduler threads are not ended if onCompleted is not called on the Observer in 1.2.7, but they are ended regardless in 2.x even if onComplete is not called, why is this so? Would it be a safe thing to end a ThreadWorker if onComplete was not called on the Observer.

In a nutshell, what is the right way to keep the number of idle RxIoScheduler threads low in 1.x?

Thanks in advance for any pointers that you can share to help us.

**Comment by akarnokd** [comment]
> Is this a normal thing to observe for the CachedThreadScheduler in 1.2.7

No. You have a scheduler leak somewhere. This could be custom operators, manual `Scheduler.createWorker` usages or never ending sources running on those schedulers.

What are you using the IO threads for?

> purge thread is turned on by default

The purge thread removes timed but cancelled tasks which Java doesn't do by default. It has nothing to do with recycling IO scheduler threads.

> Thanks in advance for any pointers that you can share to help us.

- Use `Schedulers.from` with a dedicated `ThreadPoolExecutor` where you control the maximum number of threads.
- You are running a pretty old version, the latest of 1.x is 1.3.8.
- 1.x is no longer supported and 2.x fixed many internal inconsistencies that could have lead to leaks.

**Comment by pp112018** [comment]
We use the IO threads mainly for making asynchronous service calls between services. Many times, these calls are also wrapped in Hystrix, and we provide this as a platform to hundreds of services in production. We found that many of these services are affected, but each extent varies, the severe case in thousands up to 15k. Inspection of the thread dumps show that the idle IO threads were long-lived but did very, very little CPU work.

Can you help us understand the move from setting the removal on cancel policy in 1.x to a dedicated purge thread in 2.x? What is behind the decision to using a dedicated thread in 2.x? 

I also apologize for not making clear our example. So we also surmised that there was some kind of leakage somewhere and one of the testings that we did was to observe what happens if the onCompleted call was not made in 1.x versus the onComplete call not being made in 2.x because we suspected (rightly or wrongly) that this was missing in some of the application code. Our test code is very simple, we use the Observable create factory method to create an Observable and then two Observers that subscribe to that Observable using the io scheduler. We deliberately leave out onCompleted/onComplete when creating the observable and we let the main thread idle for a few minutes to give the evictor thread more than one chance to run after the IO threads complete their work. What we saw was the IO threads forcibly being removed in 2.x but not in 1.x, that was what we meant by the "purging" in 2.x, that the behavior has changed

**Comment by pp112018** [comment]
Please ignore the last paragraph in the last comment, on diving deep into the code, we see what has happened in that test case, onCompleted was required by the SafeSubscriber in 1.x to unsubscribe which will eventually release the thread back to the queue, since this was missing in the flow created by the test code, those IO threads never get released and evicted; in 2.x, the parent Scheduler parent class wraps the work to be done within a DisposeTask that guarantees that the IO thread gets released back to the queue for eviction. The subscription list approach seems quite complex and the DisposeTask very elegant.

So, closing comments from you David, for the sake of repeating for our benefit, given how complex this one example is, there's no telling what foibles exist in our application code, the upgrade to 2.x would be best thing to do for us, correct?

Also, please don't forget to enlighten us on the design decision to use a dedicated purge thread instead of remove on cancel policy?

**Comment by akarnokd** [comment]
In general, without seeing your code, it could be both a programming error on your part or a corner case/quirk of RxJava 1.

> Can you help us understand the move from setting the removal on cancel policy in 1.x to a dedicated purge thread in 2.x? What is behind the decision to using a dedicated thread in 2.x?

It needed reflection as it was a Java 7+ feature. Certain phones mess up reflection for one. The secondary reason is that `purge` is overall more efficient and is available to Java 6+.

You can wrap an `ExecutorService` via `Schedulers.from` that has this policy enabled if you really need it.

> 2.x would be best thing to do for us, correct?

1.x is end of life so there won't be any support, changes, fixes etc. to it. 2.x has architectural, performance and API benefits over 1.x. I recommend upgrading to 2.x; most other libraries have upgraded as well by now so it shouldn't be much of a problem.

**Comment by pp112018** [comment]
We would love to share the code with you if we have even the possibility of narrowing it down to a few suspects, the problem that we face is we use RX so widely in our ecosystem and idle thread behavior is so pervasive that it is hard to pinpoint exactly what might be responsible. Some high traffic services suffer the problem more severely than others so the one thing that we are certain of is that it is load induced. Also, since we use RX at all levels and given how common the problem is, what we can reasonably surmise is that it must be some library code. However, it is hard to ascertain if it is a quirk of RX 1.2.7 at high load, or whether a bug exists in how we use 1.2.7 in either our platform or application code as you say. The only thing that is certain from looking at the 1.2.7 code is for some reason, those large number of idle threads were not released back to the queue, because for if they were, the evictor should have multiple chances to reclaim them given how long-lived the threads were.

This was what prompted us to create the abovementioned test case for we saw that the only way a ThreadWorker would be released back to the queue is if onCompleted was called on the subscriber at which unsubscription would happen; or if onError happened.

Just one minor question that I have, it seems like several layers of subscriptions can be created through the SubscriptionList and CompositeSubscription model even for our simple test case, and the code has to work through a bit of unwinding to do before it is able to get to the subscription to effectively release the worker back to the queue. We also noticed that the work to unsubscribe is submitted as a different task to the same executor that handles the work request for the thread worker unlike the DisposeTask which wraps the work request and dispose task within the same runnable. In the face of high load and multiple context switching, does the possibility of race conditions exist for 1.2.7? We have searched quite thoroughly for why onCompleted would not be called in our code and in the absence of any concrete findings, we are wondering if race conditions under high load might somewhat be responsible.

**Comment by akarnokd** [comment]
> We also noticed that the work to unsubscribe is submitted to the same executor that handles the work request for the thread worker.

This may lead to a same-pool deadlock if the scheduler is blocked on non-interruptible code, unsubscription may never run.

> In the face of high load and multiple context switching, does the possibility of race conditions exist?

No.

You could have a wrapper `Scheduler` that tracks the calls to `createWorker` and `Worker.unsubscribe` before delegating, so you could log what created those workers.

**Comment by pp112018** [comment]
> > We also noticed that the work to unsubscribe is submitted to the same executor that handles the work request for the thread worker.
> 
> This may lead to a same-pool deadlock if the scheduler is blocked on non-interruptible code, unsubscription may never run.
> 
> > In the face of high load and multiple context switching, does the possibility of race conditions exist?
> 
> No.

What makes you certain about that? We are trying to figure out why the problem is exacerbated in the face of high load
> 
> You could have a wrapper `Scheduler` that tracks the calls to `createWorker` and `Worker.unsubscribe` before delegating, so you could log what created those workers.

**Comment by akarnokd** [comment]
> What makes you certain about that?

Race condition inside the schedulers manifested themselves pretty soon during development of RxJava. If the JVM is not broken, there is no way to have a race condition within the schedulers.

> We are trying to figure out why the problem is exacerbated in the face of high load

You could do a thread dump to see what's going on. If threads are waiting in the `LockSupport.park` pretty close to a `ScheduledThreadPoolExecutor`, you don't end the sequences properly. If they are blocked somewhere in user code or external API call, you have likely non-interruptible task.

You can switch to an `ExecutorService` with a maximum number of threads and use `Schedulers.from`, but note that this will eventually hang if all threads get blocked an never unblock.

**Comment by pp112018** [comment]
> > What makes you certain about that?
> 
> Race condition inside the schedulers manifested themselves pretty soon during development of RxJava. If the JVM is not broken, there is no way to have a race condition within the schedulers.
> 
> > We are trying to figure out why the problem is exacerbated in the face of high load
> 
> You could do a thread dump to see what's going on. If threads are waiting in the `LockSupport.park` pretty close to a `ScheduledThreadPoolExecutor`, you don't end the sequences properly. If they are blocked somewhere in user code or external API call, you have likely non-interruptible task.
> 
> You can switch to an `ExecutorService` with a maximum number of threads and use `Schedulers.from`, but note that this will eventually hang if all threads get blocked an never unblock.

We did that, they are all waiting in LockSupport.park which made us contrive the test case that did not call onCompleted in 1.2.7. We were also able to replicate this in 1.3.8 but the problem disappears when we update the test to use 2.2.3. 

Many thanks for your time, I think that in the face of so many uncertain probable root causes in 1.x against the guarantees that we saw with the DisposeTask in 2.2.3, we really have no option but to upgrade.

Please feel free to close this issue. Once again, many thanks!
